design/use-statement.md

```diff
@@ -0,0 +1,162 @@
+# The `use` statement
+
+There are currently 9 different [tool or installer](https://docs.microsoft.com/en-us/azure/devops/pipelines/tasks/tool/go-tool?view=vsts) tasks.
+Conceptually they do roughly the same thing (in the user's mind):
+- Get a known version of some tool, maybe from the internet
+- Set up the CI environment so that when I invoke `tool`, I get that version.
+
+## Problems
+
+### Semantics of the task
+Some tools are hard to install on the fly.
+So, these are really "use a version from some limited subset of possible versions".
+Others can be used to fetch arbitrary versions from the internet.
+We introduced a split between "use version" and "install tool", but this has turned out to be more confusing than helpful in practice.
+
+### Naming
+The tasks are not named consistently, and they take inconsistent inputs.
+We have 2 `*Tool` tasks, 3 `*Installer` tasks, 2 `*ToolInstaller` tasks, and 2 `Use*Version` tasks.
+Some take a `version`, others a `versionSpec`, and one a `versionSelector`.
+See the appendix for a list of current tasks.
+
+### YAML syntax
+Some of the tasks take complex, interdependent inputs (Visual Studio Test Platform Installer).
+This is really hard to work with in YAML.
+Also, our `taskName@taskVersion` syntax adds a layer of confusion.
+```yaml
+- task: UsePythonVersion@0
+  inputs:
+    versionSpec: 3.x
+```
+The tasks are all at version 0 right now, so real-world confusion is *probably* limited.
+
+### Other things we want to include
+- Proxy setup.
+The agent knows about its proxy; the tasks don't always pass that along to client tooling.
+- Auth setup.
+If I'm going to publish to a private npm registry, I should be able to give the creds here.
+- Problem matchers.
+If the Python compiler emits errors in a known format, we should be able to read that via regex and surfaec it in our UI.
+VS Code has a defined format for problem matchers.
+
+## Solution
+
+We'll add new YAML syntax which puts tools on the `PATH`, adds proxy information, adds problem matchers, and optionally configures auth.
+This syntax will be powered by tasks, giving a familiar extensibility point for third parties to work with.
+
+### New syntax
+
+YAML will get a new syntax for `use`-ing a tool or ecosystem.
+```yaml
+- use: {toolName} {versionSpec}
+```
+
+This can be implemented as sugar over the existing syntax, much like the `powershell` and `bash` keywords today.
+
+`{versionSpec}` is optional, and not needed in some cases (see below).
+It can also be passed as an explicit parameter, like this:
+```yaml
+- use: {toolName}
+  version: {versionSpec}
+```
+
+By default, `use` will set up the target ecosystem to use the agent's proxy.
+If this behavior isn't desired, it can be overridden with:
+```yaml
+- use: someTool someVersion
+  proxy: false
+```
+
+If a tool needs credentials, those can be added like this:
+```yaml
+resources:
+  feeds:
+  - feed: myFeed
+    auth: {auth_information}
+
+steps:
+- use: someTool someVersion
```

Comment:
Do we anticipate any version specs having spaces. I wouldnt think so, but i want to make sure that if we support this inline-version syntax, we keep the parser really really dumb. Similarly for the tool portion, do we support spaces for the tool name? If so, we should just not do this inline-version syntax.